File: src/services/mentionService.js
const User = require('../models/User');
const Notification = require('../models/Notification');
const { emitToUser } = require('./socketService');

/**
 * Parse @mentions from text content
 * Returns array of mentioned usernames
 */
const parseMentions = (text) => {
  if (!text) return [];
  const mentionRegex = /@(\w+(?:\.\w+)*)/g;
  const mentions = [];
  let match;
  while ((match = mentionRegex.exec(text)) !== null) {
    mentions.push(match[1]);
  }
  return [...new Set(mentions)];
};

/**
 * Resolve usernames to user IDs
 */
const resolveMentions = async (usernames) => {
  if (!usernames || usernames.length === 0) return [];

  // Build regex for exact word matching
  const patterns = usernames.map((name) => new RegExp(`^${name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}$`, 'i'));

  const users = await User.find({
    $or: [
      { name: { $in: patterns } },
      { email: { $in: patterns } },
    ],
  }).select('_id name email');

  return users;
};

/**
 * Send mention notifications to users
 */
const notifyMentions = async (mentionedUsers, { entityType, entityId, entityTitle, actorId, actorName }) => {
  for (const user of mentionedUsers) {
    if (user._id.toString() === actorId.toString()) continue; // Don't notify yourself

    const notification = await Notification.create({
      user: user._id,
      type: 'COMMENT_ADDED', // Reuse existing type for mentions
      message: `${actorName} mentioned you in "${entityTitle}"`,
      relatedEntity: { type: entityType, id: entityId },
    });

    // Real-time push via Socket.IO
    emitToUser(user._id, 'notification', {
      type: 'MENTION',
      message: notification.message,
      entityType,
      entityId,
      timestamp: notification.createdAt,
    });
  }
};

module.exports = { parseMentions, resolveMentions, notifyMentions };

File: backend/src/main/java/com/taskflow/service/MentionService.java
package com.taskflow.service;

import com.taskflow.entity.Notification;
import com.taskflow.entity.User;
import com.taskflow.entity.enums.NotificationType;
import com.taskflow.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Parses and resolves {@code @name} mentions and notifies mentioned users. */
@Service
public class MentionService {

    private static final Pattern MENTION_PATTERN = Pattern.compile("@(\\w+(?:\\.\\w+)*)");

    private final UserRepository userRepository;
    private final NotificationService notificationService;

    public MentionService(UserRepository userRepository, NotificationService notificationService) {
        this.userRepository = userRepository;
        this.notificationService = notificationService;
    }

    public List<String> parseMentions(String text) {
        if (text == null || text.isBlank()) {
            return List.of();
        }
        LinkedHashSet<String> mentions = new LinkedHashSet<>();
        Matcher matcher = MENTION_PATTERN.matcher(text);
        while (matcher.find()) {
            mentions.add(matcher.group(1));
        }
        return new ArrayList<>(mentions);
    }

    public List<User> resolveMentions(List<String> usernames) {
        if (usernames == null || usernames.isEmpty()) {
            return List.of();
        }
        List<String> lowered = usernames.stream()
                .filter(name -> name != null && !name.isBlank())
                .map(name -> name.toLowerCase())
                .distinct()
                .toList();
        if (lowered.isEmpty()) {
            return List.of();
        }
        return userRepository.findByLowerNameOrLowerEmailIn(lowered);
    }

    public void notifyMentions(List<User> mentionedUsers, Long actorId, String actorName,
                               String entityType, Long entityId, String entityTitle) {
        for (User user : mentionedUsers) {
            if (user.getId().equals(actorId)) {
                continue;
            }
            String message = actorName + " mentioned you in \"" + entityTitle + "\"";
            Notification notification = notificationService.create(user, NotificationType.COMMENT_ADDED,
                    message, entityType, entityId);
            notificationService.push(user.getId(), "MENTION", message, entityType, entityId,
                    notification.getCreatedAt());
        }
    }
}
